### ISSUE TITLE
CMake custom_targets when used as a subdirectory

### ISSUE THREAD

**Issue by sphaero** [zeromq/zproject#1237]
including multiple cmake projects as subdirectories in another cmake project fails because the custom_targets need to be unique.

```
CMake Error at ext/libsphactor/CMakeLists.txt:390 (add_custom_target):
  add_custom_target cannot create target "copy-selftest-ro" because another
  target with the same name already exists.  The existing target is a custom
  target created in source directory
  "/home/arnaud/src/gazebosc-master/ext/czmq".  See documentation for policy
  CMP0002 for more details.


CMake Error at ext/libsphactor/CMakeLists.txt:395 (add_custom_target):
  add_custom_target cannot create target "make-selftest-rw" because another
  target with the same name already exists.  The existing target is a custom
  target created in source directory
  "/home/arnaud/src/gazebosc-master/ext/czmq".  See documentation for policy
  CMP0002 for more details.

...
```
A fix could be to check whether they exist already? https://public.kitware.com/pipermail/cmake/2011-July/045269.html

This is perhaps a bit of an edge case?

**Comment by sphaero** [comment]
I can confirm that the suggested fix works. However I can't oversee the consequences fully. To me it seems the fix won't hurt as it only skips target creation if they already exist which won't be true unless you're using multiple projects. Which is what this is about.

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 80ca23f..97696bb 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -387,15 +387,19 @@ set(TEST_CLASSES
 )
 
 
-add_custom_target(
-    copy-selftest-ro ALL
-    COMMAND ${CMAKE_COMMAND} -E copy_directory ${PROJECT_SOURCE_DIR}/src/selftest-ro ${PROJECT_BINARY_DIR}/src/selftest-ro
-)
+if (NOT TARGET copy-selftest-ro)
+    add_custom_target(
+        copy-selftest-ro ALL
+        COMMAND ${CMAKE_COMMAND} -E copy_directory ${PROJECT_SOURCE_DIR}/src/selftest-ro ${PROJECT_BINARY_DIR}/src/selftest-ro
+    )
+endif()
 
+if (NOT TARGET make-selftest-rw)
 add_custom_target(
     make-selftest-rw ALL
     COMMAND ${CMAKE_COMMAND} -E make_directory ${PROJECT_BINARY_DIR}/src/selftest-rw
 )
+endif()
 
 set_directory_properties(
     PROPERTIES
@@ -422,7 +426,9 @@ include(CTest)
 ########################################################################
 # cleanup
 ########################################################################
+if (NOT TARGET distclean)
 add_custom_target (distclean @echo Cleaning for source distribution)
+endif()
 
 set(cmake_generated ${PROJECT_BINARY_DIR}/CMakeCache.txt
                     ${PROJECT_BINARY_DIR}/cmake_install.cmake
diff --git a/builds/cmake/Modules/ClangFormat.cmake b/builds/cmake/Modules/ClangFormat.cmake
index 032beb3..7f5a2ca 100644
--- a/builds/cmake/Modules/ClangFormat.cmake
+++ b/builds/cmake/Modules/ClangFormat.cmake
@@ -18,10 +18,12 @@ if("${CLANG_FORMAT}" STREQUAL "")
   set(CLANG_FORMAT "clang-format")
 endif()
 
+if (NOT TARGET clang-format)
 add_custom_target(
         clang-format
         COMMAND ${CLANG_FORMAT} -style=file -i ${ALL_SOURCE_FILES}
 )
+endif()
 
 function(JOIN VALUES GLUE OUTPUT)
   string (REPLACE ";" "${GLUE}" _TMP_STR "${VALUES}")
@@ -30,23 +32,29 @@ endfunction()
 
 configure_file(builds/cmake/clang-format-check.sh.in clang-format-check.sh @ONLY)
 
+if (NOT TARGET clang-format-check)
 add_custom_target(
         clang-format-check
         COMMAND chmod +x clang-format-check.sh
         COMMAND ./clang-format-check.sh
         COMMENT "Checking correct formatting according to .clang-format file using ${CLANG_FORMAT}"
 )
+endif()
 
+if (NOT TARGET clang-format-check-CI)
 add_custom_target(
         clang-format-check-CI
         COMMAND chmod +x clang-format-check.sh
         COMMAND ./clang-format-check.sh --CI
         COMMENT "Checking correct formatting according to .clang-format file using ${CLANG_FORMAT}"
 )
+endif()
 
+if (NOT TARGET clang-format-diff)
 add_custom_target(
         clang-format-diff
         COMMAND ${CLANG_FORMAT} -style=file -i ${ALL_SOURCE_FILES}
         COMMAND git diff ${ALL_SOURCE_FILES}
         COMMENT "Formatting with clang-format (using ${CLANG_FORMAT}) and showing differences with latest commit"
 )
+endif()
```

**Comment by sphaero** [comment]
created a PR: https://github.com/zeromq/zproject/pull/1238

**Comment by sphaero** [comment]
closed by #1238
